torch/lib/THD/base/data_channels/DataChannelNccl.hpp

```diff
@@ -0,0 +1,232 @@
+#pragma once
+
+#include "../DataChannel.hpp"
+#include "DataChannelUtils.hpp"
+
+#include <nccl.h>
+
+#include <utility>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+
+#define NCCL_CHECK(cmd) do {                                  \
+  ncclResult_t error = cmd;                                   \
+  if (error != ncclSuccess) {                                 \
+    std::string err = "NCCL error in: " +                     \
+                      std::string(__FILE__) + ":" +           \
+                      std::to_string(__LINE__) + ", " +       \
+                      std::string(ncclGetErrorString(error)); \
+    throw std::runtime_error(err);                            \
+  }                                                           \
+} while (0)
+
+
+namespace thd {
+
+// Type aliasing
+using NcclResourcePair =
+  std::pair<std::vector<ncclComm_t>*, std::vector<cudaEvent_t>*>;
+
+struct DataChannelNccl : DataChannel {
+
+  // Nothing to implement
+  struct RequestNccl : DataChannel::Request {};
+
+  // Wrapper on the pair of NCCL resources
+  class NcclResources {
+
+  public:
+
+    NcclResources() = default;
+    NcclResources(std::unique_ptr<std::vector<ncclComm_t>>&& ncclComm,
+                  std::unique_ptr<std::vector<cudaEvent_t>>&& event):
+
+      _commEventPair(std::pair<std::unique_ptr<std::vector<ncclComm_t>>,
+                               std::unique_ptr<std::vector<cudaEvent_t>>>
+                               (std::move(ncclComm), std::move(event))) {}
+    // Delete copy and assignment ctors
+    NcclResources(const NcclResources&) = delete;
+    NcclResources& operator=(const NcclResources&) = delete;
+
+    // Move ctors by default
+    NcclResources(NcclResources&&) = default;
+    NcclResources& operator=(NcclResources&&) = default;
+
+    // Nccl Communicator Getter
+    std::vector<ncclComm_t>* ncclComms() {
+      return _commEventPair.first.get();
+    }
+
+    // Nccl CUDA event Getter
+    std::vector<cudaEvent_t>* ncclCudaEvents() {
+      return _commEventPair.second.get();
+    }
+
+  private:
+
+    std::pair<std::unique_ptr<std::vector<ncclComm_t>>,
+              std::unique_ptr<std::vector<cudaEvent_t>>> _commEventPair;
```

Comment:
They will always be very small, so any costs you could possibly pay because of missing a move (I'm also looking for missed moves when reviewing), are probably outweighed by the cost of double indirection when accessing memory all the time. Storing them as two members is going to make the code a lot clearer - if you didn't make them private, you could just access them like regular fields, and wouldn't need `NcclResourcePair` at all - you could just pass `NcclResources&` or `NcclResources*` around and have everything work just fine.